WordPress/openverse-catalog
# iNaturalist in-SQL loading

## Related to
This relates to #689 in that my understanding is that we have not yet completed a full load of iNaturalist data, and if we were going to start loading the data incrementally, we would probably want a full load to start from. In practice, it also runs the cleaning and upsert steps on smaller batches, rather than all at once across 120+ million records.

It also integrates Catalog of Life data to address #696.

## Description
The existing iNaturalist dag essentially follows these steps: load normalized data from s3 to PostgreSQL, pull integrated JSON out in batches of 10,000 joined records, and then follow the rest of the steps in any provider dag, to reload the JSON into PostgreSQL.

The initial load of normalized data takes on the order of 10-15 minutes for approximately 120,000,000 records in `inaturalist.photos', and somewhat fewer in the other inaturalist tables. However, the dag has timed out after several days on the step where joined the data is being pulled out of PostgreSQL. 

This PR moves all of the processing to SQL. The latest local run timed out after 24 hours and 87,943,312 records loaded to `public.image`. I increased the timeout to be 48 hours, but/and we could consider longer. I estimate that it should complete within 36 hours. More on that [on the chart 12-14-2022 tab in this google sheet](https://docs.google.com/spreadsheets/d/1gaVsvFnsYby2iwzRm0Ta9IPhvNixCpQdHByGLo1r_pg/edit#gid=108859884). The workbook also contains visualizations for other test runs during this development process.

### Issues I will make and add to comments for future work

- Improve Airflow reporting, right now it only reports the number of records loaded
- Consider adding indices to the temp loading table _after_ loading data into it.
- Figure out a better way to store batch limit, for DRYer code
- Keep copies of the most recent source files on s3.

## Testing Instructions
I increased the disk image size in my local docker to 192 GB. You should be able to run the whole thing with a very small sample of test data in the repo, and I made some changes to `tests/dags/providers/provider_api_scripts/resources/inaturalist/pull_sample_records.py` to make it easier/faster to pull a test sample of your choice. And then poke around for data quality/formatting stuff. [Instructions to download the inaturalist data.](https://github.com/inaturalist/inaturalist-open-data/tree/main/Metadata/Download) 

I could really use a second set of eyes to double check that I have implemented all of the python data cleaning steps (initial attempt [on the "image table ddl" tab here](https://docs.google.com/spreadsheets/d/1gaVsvFnsYby2iwzRm0Ta9IPhvNixCpQdHByGLo1r_pg/edit#gid=2082506887), in the SQL. After running the large majority of the dataset, the log for the python cleaning step looks promising to me, but without more on #898, a lot of this has to happen manually. 




## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
